### DISCUSSION TITLE
Can't use moveUp or moveDown function

### DISCUSSION THREAD

**Discussion by DungnvKaopiz** [MathKeyboardEngine/MathKeyboardEngine#10] [Q&A] [answered]
Hi, i'm having problem with moveUp and moveDown function. I tried to debug it and found out my keyboardMemory object doesn't have width ?
Here's the code that i try to debug with:

` window.addEventListener("keydown", (e) => {
if (keyCode == 38) {
              activeKey(mke.moveUp); )}`

```
const activeKey = async (onclickArrowFuncForKeyboardMemory) => 
 {
          onclickArrowFuncForKeyboardMemory(keyboardMemory);

          let fromPlaceholder =
            keyboardMemory.current instanceof mke.Placeholder
              ? keyboardMemory.current
              : keyboardMemory.current.parentPlaceholder;
          let suggestingNode = fromPlaceholder.parentNode;

          getMoveUpSuggestion(fromPlaceholder, suggestingNode);
          displayResult();
};
```

```
        const getMoveUpSuggestion = (fromPlaceholder, suggestingNode) => {
          const { rowIndex, columnIndex } = getPositionOf(
            fromPlaceholder,
            suggestingNode
          );
          if (rowIndex - 1 >= 0) {
            return suggestingNode.grid[rowIndex - 1][columnIndex];
          } else {
            return null;
          }
        };
```
```
        const getPositionOf = (placeholder, suggestingNode) => {
          const index = suggestingNode.placeholders.indexOf(placeholder);
          if (index == -1) {
            throw "The provided Placeholder is not part of this MatrixNode.";
          }
          const rowIndex = Math.floor(index / suggestingNode.width);
          const columnIndex = index - rowIndex * suggestingNode.width;
          console.log(keyboardMemory);
          console.log(suggestingNode.width);
          return { rowIndex, columnIndex };
        };
```

and here's what show in console:
![image](https://github.com/MathKeyboardEngine/MathKeyboardEngine/assets/85537564/05900a97-2d61-4b94-b131-302db2cf1026)

Any idea what i've been doing wrong ? thanks for reading

**Comment by SymboLinker** [comment]
Hi,

Just for other readers: there is no need to use `getMoveUpSuggestion`, `getPositionOf ` or a `fromPlaceholder ` in your code. (@DungnvKaopiz has added those methods for debugging purposes.) Just using `mke.moveUp(keyboardMemory)` should suffice, because the method `moveUp` is designed to find out by itself whether a move up is possible. Something like this should work:
```
 window.addEventListener("keydown", (e) => { 
  if (e.keyCode == 38) { 
    mke.moveUp(myKeyboardMemory);
    myDisplayResultMethod();
  }
});
```
<br />

> Any idea what i've been doing wrong?

The methods `getMoveUpSuggestion` and `getPositionOf` that your use, are the methods from a [MatrixNode](https://github.com/MathKeyboardEngine/MathKeyboardEngine/blob/main/src/SyntaxTreeComponents/Nodes/BranchingNodes/MatrixNode.ts). In other words, those are wrong here.

Are you using a [DescendingBranchingNode](https://github.com/MathKeyboardEngine/MathKeyboardEngine/blob/main/src/SyntaxTreeComponents/Nodes/BranchingNodes/DescendingBranchingNode.ts) when inserting a fraction?

In your screenshot, the `keyboardMemory.currentNode` is a `DigitNode` with value "5". Its parent is "the second placeholder of the fraction node". That fraction is doing the so-called "suggestion", so the suggesting node could in this case be found via:
```
let denominator = keyboardMemory.current.parentPlaceholder;
let fraction = denominator.parentNode; // this is the "suggesting node"
```
Is this suggesting node a `DescendingBranchingNode`?
<br />

> I tried to debug it and found out my keyboardMemory object doesn't have width?

That is correct. See the [KeyboardMemory class definition](https://github.com/MathKeyboardEngine/MathKeyboardEngine/blob/main/src/KeyboardEngine/KeyboardMemory.ts). 

You probably mean that your `suggestingNode` does not have width. A `MatrixNode` has a (grid)width, but a `DescendingBranchingNode` does not have a width.

**Reply by DungnvKaopiz** [reply]
Thanks for answering and sorry for the long reply, stuff happended :)
And you're right, i was debugging using MatrixNode code while using DescendingBranchingNode.
So i try debugging with the DescendingBranchingNode code and it works! So the real problem here is why when i use
```
 window.addEventListener("keydown", (e) => { 
  if (e.keyCode == 38) { 
    mke.moveUp(myKeyboardMemory);
    myDisplayResultMethod();
  }
});
```
It does not work. Keep in mind that my delete Left, move left, move right, ... are still working perfectly fine

**Reply by DungnvKaopiz** [reply, marked as answer]
ok, looks like i know what the problem here:
here's my code
![image](https://github.com/MathKeyboardEngine/MathKeyboardEngine/assets/85537564/4263fe21-1f0c-4e77-89ba-cc6f561905ea)

and here's the doc:
![image](https://github.com/MathKeyboardEngine/MathKeyboardEngine/assets/85537564/805d8d94-9f1a-478b-a691-12113e6dac76)

so i just need to switch StandardBranchingNode to DescendingBranchingNode

**Reply by SymboLinker** [reply]
Good, very nice that you have found the solution and thanks for sharing :)
